vsutil.py

```diff
@@ -176,3 +195,75 @@ def is_image(filename: str, /) -> bool:
     Returns true if a filename refers to an image.
     """
     return types_map.get(path.splitext(filename)[-1], '').startswith('image/')
+
+
+def depth(clip: vs.VideoNode,
+          bitdepth: int,
+          /,
+          sample_type: Optional[Union[int, vs.SampleType]] = None,
+          *,
+          range: Optional[Union[int, Range]] = None,
+          range_in: Optional[Union[int, Range]] = None,
+          dither_type: Optional[Union[Dither, str]] = None) \
+        -> vs.VideoNode:
+    """
+    A bit depth converter only using core.resize and Format.replace.
+    By default, outputs FLOAT sample type for 32 bit and INTEGER for anything else.
+
+    :param bitdepth:    Desired bits_per_sample of output clip.
+    :param sample_type: Desired sample_type of output clip. Allows overriding default FLOAT/INTEGER behavior. Accepts
+                        vapoursynth.SampleType enums INTEGER and FLOAT or their values, [0, 1].
+    :param range:       Output pixel range (defaults to input clip's range). See `Range`.
+    :param range_in:    Input pixel range (defaults to input clip's range). See `Range`.
+    :param dither_type: Dithering algorithm. Allows overriding default dithering behavior. See `Dither`.
+        Defaults to Floyd-Steinberg error diffusion when downsampling, converting between ranges, or upsampling full
+        range input. Defaults to 'none', or round to nearest, otherwise.
+
+    :return: Converted clip with desired bit depth and sample type. ColorFamily will be same as input.
+    """
+    def readable_enums(enum, module: str = 'vsutil') -> list:
+        """
+        Returns a readable error message listing all possible values in `module.enum`.
+        """
+        return [module + "." + str(i) for i in enum] + [i.value for i in enum]
+
+    # enum mapping and error checking
+    if sample_type is not None:
+        try:
+            sample_type = vs.SampleType(sample_type)
+        except ValueError:
+            raise ValueError(f'depth: sample_type must be in {readable_enums(vs.SampleType, str(vs.__name__))}.') from None
+    if range is not None:
+        try:
+            range = Range(range)
+        except ValueError:
+            raise ValueError(f'depth: range must be in {readable_enums(Range)}.') from None
+    if range_in is not None:
+        try:
+            range_in = Range(range_in)
+        except ValueError:
+            raise ValueError(f'depth: range_in must be in {readable_enums(Range)}.') from None
+    if dither_type is not None:
+        try:
+            dither_type = Dither(dither_type)
+        except ValueError:
+            raise ValueError(f'depth: dither_type must be in {readable_enums(Dither)}.') from None
+
```

Comment:
We only have to write this code once.
I’m thinking of something like
```py
def resolve_enum(enum_type, value):
    try:
        return enum_type(value)
    except ValueError:
        raise ValueError(f'value must be in {[f"{module}.{str(e)} ({e.value})" for e in enum_type]}') from None
```
which would look like
```py
In [22]: resolve_enum(Range, 0)
Out[22]: <Range.LIMITED: 0>

In [23]: resolve_enum(Range, 1)
Out[23]: <Range.FULL: 1>

In [24]: resolve_enum(Range, 2)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-24-7a2f94c717b0> in <module>
----> 1 resolve_enum(Range, 2)

<ipython-input-17-e5bc1e1b3f60> in resolve_enum(enum_type, value)
      3         return enum_type(value)
      4     except ValueError:
----> 5         raise ValueError(f'value must be in {[f"{module}.{str(e)} ({e.value})" for e in enum_type]}') from None
      6

ValueError: value must be in ['vsutil.Range.LIMITED (0)', 'vsutil.Range.FULL (1)']
```

Reply:
If accessing generic vars from types wasn't such a hassle, I would suggest getting the type from the function's `__annotations__` even, but this would be less readable.

Reply:
Why not manually override `__repr__`?

Something along the lines of:

```py
    def __repr__(self):
        return f"<{self.__module__}.{self.__qualname__}.{self.name}: {self.value}>"
```

Reply:
Addressed in 4e2915a5f27080fbb1366c1607f8c8fc912058a5. Let me know if you have any suggestions for the two functions I added, I also added small test cases for them. I think _resolve_enum satisfies what kageru suggested. Keeping _readable_enums as its own function for now as it might be useful in the future separately from the _resolve_enum use.